cas-server-webapp/src/main/java/org/jasig/cas/web/support/AbstractThrottledSubmissionHandlerInterceptorAdapter.java

```diff
@@ -0,0 +1,197 @@
+
+    /** The Logger instance. */
+    protected final Logger logger = LoggerFactory.getLogger(getClass());
+
+    @Min(0)
+    private int failureThreshold = DEFAULT_FAILURE_THRESHOLD;
+
+    @Min(0)
+    private int failureRangeInSeconds = DEFAULT_FAILURE_RANGE_IN_SECONDS;
+
+    @NotNull
+    private String usernameParameter = DEFAULT_USERNAME_PARAMETER;
+
+    private double thresholdRate;
+
+    @Override
+    public void afterPropertiesSet() throws Exception {
+        this.thresholdRate = (double) failureThreshold / (double) failureRangeInSeconds;
+    }
+
+    @Override
+    public final boolean preHandle(final HttpServletRequest request, final HttpServletResponse response, final Object o)
+            throws Exception {
+        if (!HttpMethod.POST.name().equals(request.getMethod())) {
+            return true;
+        }
+
+        if (exceedsThreshold(request)) {
+            recordThrottle(request);
+            request.setAttribute(WebUtils.CAS_ACCESS_DENIED_REASON, "screen.blocked.message");
+            response.sendError(HttpStatus.SC_FORBIDDEN,
+                    "Access Denied for user [" + request.getParameter(usernameParameter)
+                            + "] from IP Address [" + request.getRemoteAddr() + ']');
+            return false;
+        }
+
+        return true;
+    }
+
+    @Override
+    public final void postHandle(
+            final HttpServletRequest request,
+            final HttpServletResponse response,
+            final Object o, final ModelAndView modelAndView
+    ) throws Exception {
+        if (!HttpMethod.POST.name().equals(request.getMethod())) {
+            return;
+        }
+
+        final RequestContext context = (RequestContext) request.getAttribute("flowRequestContext");
+
+        if (context == null || context.getCurrentEvent() == null) {
+            return;
+        }
+
+        // retrieve the Login Context, where `.getFlowScope()` may not be available
+        final String loginContext;
+        try {
+            loginContext = (String) context.getFlowScope().get("jsonLoginContext");
+        } catch (final IllegalStateException e) {
+            logger.warn(e.getMessage());
+            return;
+        }
+        // no need to catch json parsing exceptions, since if `loginContext` exists, it must be good
+        final OpenScienceFrameworkLoginHandler.OpenScienceFrameworkLoginContext osfLoginContext
+                = OpenScienceFrameworkLoginHandler.OpenScienceFrameworkLoginContext.fromJson(loginContext);
+        if (osfLoginContext == null || osfLoginContext.getHandleErrorName() == null) {
+            logger.warn("Fail to retrieve login context and authentication exceptions.");
+            return;
+        }
+        if (OpenScienceFrameworkAuthenticationExceptionHandler.isTriggerThrottleIncrease(osfLoginContext.getHandleErrorName())) {
+            // record submission failure only for throttle-trigger auth errors
+            recordSubmissionFailure(request);
```

Comment:
Record submission failure only when one of three trigger-exceptions exist in `loginContext`.